lib/net.js

```diff
@@ -323,12 +326,16 @@ Socket.prototype._onTimeout = function() {
 
 Socket.prototype.setNoDelay = function(enable) {
   // backwards compatibility: assume true when `enable` is omitted
+  this._setNoDelay = enable === undefined ? true : !!enable;
+
   if (this._handle && this._handle.setNoDelay)
-    this._handle.setNoDelay(enable === undefined ? true : !!enable);
+    this._handle.setNoDelay(this._setNoDelay);
 };
```

Comment:
I don't think you need new flags for this:

```javascript
Socket.prototype.setNoDelay = function(enable) {
  if (this._connected === false) {
    this.once('connect', this.setNoDelay.bind(this, enable));
    return;
  }
  // ...
};
```

The use of `Function#bind()` is intentional, it avoids the unconditional closure allocation that would happen if you wrote the uncommon path as `this.once('connect', function() { this.setNoDelay(enable); })`.